runtime/oti/j9accessbarrier.h

```diff
@@ -104,24 +104,16 @@ typedef struct J9IndexableObject* mm_j9array_t;
 #define J9_CONVERT_POINTER_TO_TOKEN__(vmThread, pointer) J9_CONVERT_POINTER_TO_TOKEN_VM__(J9VMTHREAD_JAVAVM(vmThread),pointer)
 
 #define J9JAVAARRAYDISCONTIGUOUS_EA(vmThread, array, index, elemType) \
-	(J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmThread) \
-		? (&(((elemType*)J9_CONVERT_POINTER_FROM_TOKEN__(vmThread, ((U_32*)((J9IndexableObjectDiscontiguousCompressed *)array + 1))[((U_32)index)/(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize/sizeof(elemType))])) \
-		: (&(((elemType*)(((UDATA*)((J9IndexableObjectDiscontiguousFull *)array + 1))[((U_32)index)/(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize/sizeof(elemType))])))
+	(&(((elemType*)(((UDATA*)((UDATA)array + vmThread->javaVM->discontiguousHeaderSize))[((U_32)index)/(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize/sizeof(elemType))]))
 
 #define J9JAVAARRAYDISCONTIGUOUS_EA_VM(javaVM, array, index, elemType) \
-	(J9JAVAVM_COMPRESS_OBJECT_REFERENCES(javaVM) \
-		? (&(((elemType*)J9_CONVERT_POINTER_FROM_TOKEN_VM__(javaVM, ((U_32*)((J9IndexableObjectDiscontiguousCompressed *)array + 1))[((U_32)index)/((javaVM)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%((javaVM)->arrayletLeafSize/sizeof(elemType))])) \
-		: (&(((elemType*)(((UDATA*)((J9IndexableObjectDiscontiguousFull *)array + 1))[((U_32)index)/((javaVM)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%((javaVM)->arrayletLeafSize/sizeof(elemType))])))
+	(&(((elemType*)(((UDATA*)((UDATA)array + javaVM->discontiguousHeaderSize))[((U_32)index)/((javaVM)->arrayletLeafSize/sizeof(elemType))]))[((U_32)index)%((javaVM)->arrayletLeafSize/sizeof(elemType))]))
 
 #define J9JAVAARRAYCONTIGUOUS_EA(vmThread, array, index, elemType) \
-	(J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmThread) \
-		? (&((elemType*)((((J9IndexableObjectContiguousCompressed *)(array)) + 1)))[(index)]) \
-		: (&((elemType*)((((J9IndexableObjectContiguousFull *)(array)) + 1)))[(index)]))
+	(&((elemType*)((UDATA*)((UDATA)array + vmThread->javaVM->contiguousHeaderSize)))[(index)])
```

Comment:
We probably want to cache (dis)contiguousHeaderSize in VMThread. This API variant is way more frequently used than the _VM variant. When we introduce offheap we will be querying isDataAddressPresent as well, so we may cache that one too.